lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -0,0 +1,730 @@
+  for (unsigned i = 0, e = apply->getNumOperands() - 1; i != e; ++i) {
+    // If any of the arguments is a non-constant value, then we can't fold this
+    // call.
+    auto op = apply->getOperand(i + 1);
+    SymbolicValue argValue = getConstantValue(op);
+    if (!argValue.isConstant())
+      return argValue;
+    paramConstants.push_back(argValue);
+  }
+
+  // TODO(constexpr patch): This is currently unused, so we don't need to
+  // calculate the correct value. Eventually, include code that calculates the
+  // correct value.
+  SubstitutionMap calleeSubMap;
+
+  // Now that have successfully folded all of the parameters, we can evaluate
+  // the call.
+  evaluator.pushCallStack(apply->getLoc().getSourceLoc());
+  SmallVector<SymbolicValue, 4> results;
+  auto callResult = evaluateAndCacheCall(*callee, calleeSubMap, paramConstants,
+                                         results, numInstEvaluated, evaluator);
+  evaluator.popCallStack();
+  if (callResult.hasValue())
+    return callResult.getValue();
+
+  unsigned nextResult = 0;
+
+  // If evaluation was successful, remember the results we captured in our
+  // current function's state.
+  if (unsigned numNormalResults = conventions.getNumDirectSILResults()) {
+    // TODO: unclear when this happens, is this for tuple result values?
+    assert(numNormalResults == 1 && "Multiple results aren't supported?");
+    setValue(apply->getResults()[0], results[nextResult]);
+    ++nextResult;
+  }
+
+  assert(nextResult == results.size() && "Unexpected number of results found");
+
+  // We have successfully folded this call!
+  return None;
+}
+
+/// Return the SymbolicValue for the specified SIL value, lazily computing
+/// it if needed.
+SymbolicValue ConstExprFunctionState::getConstantValue(SILValue value) {
+  // Check to see if we already have an answer.
+  auto it = calculatedValues.find(value);
+  if (it != calculatedValues.end())
+    return it->second;
+
+  // Compute the value of a normal instruction based on its operands.
+  auto result = computeConstantValue(value);
+
+  // If this is the top-level lazy interpreter, output a debug trace.
+  if (!fn) {
+    LLVM_DEBUG(llvm::dbgs() << "ConstExpr top level: "; value->dump());
+    LLVM_DEBUG(llvm::dbgs() << "  RESULT: "; result.dump());
+  }
+
+  setValue(value, result);
+  return result;
+}
+
+/// Evaluate the specified instruction in a flow sensitive way, for use by
+/// the constexpr function evaluator.  This does not handle control flow
+/// statements.  This returns None on success, and an Unknown SymbolicValue with
+/// information about an error on failure.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::evaluateFlowSensitive(SILInstruction *inst) {
+  // These are just markers.
+  if (isa<DebugValueInst>(inst) || isa<DebugValueAddrInst>(inst) ||
+      isa<EndAccessInst>(inst) ||
+      // Constant have no important state.
+      isa<DestroyAddrInst>(inst) || isa<RetainValueInst>(inst) ||
```

Comment:
@marcrasi  Can we just say in comments that "the interpreter doesn't model these memory management instructions and hence skips them.", instead of "Constant have no important state"